iree/compiler/Codegen/SPIRV/KernelConfig.cpp

```diff
@@ -319,24 +320,87 @@ static LogicalResult setOpConfig(spirv::ResourceLimitsAttr limits,
     std::array<int64_t, 3> workgroupSize = {1, 1, 1};
-    auto funcOp = op->getParentOfType<FuncOp>();
     return setOpConfigAndEntryPointFnTranslation(funcOp, op, {}, {}, pipeline,
                                                  workgroupSize);
   }
 
-  const int64_t subgroupSize = limits.subgroup_size().getValue().getSExtValue();
-  int64_t numElementsPerWorkgroup = subgroupSize;
-  int64_t numElementsPerThread = 1;
-  auto pipeline = IREE::Codegen::DispatchLoweringPassPipeline::SPIRVDistribute;
+  const int subgroupSize = limits.subgroup_size().getValue().getSExtValue();
+  const unsigned loopDepth = partitionedLoops.back() + 1;
+
+  // Configurations we need to decide.
+  std::array<int64_t, 3> workgroupSize;
+  SmallVector<int64_t> workgroupTileSizes;
+  SmallVector<int64_t> threadTileSizes;
+
+  // Initialize the configuration.
+  auto initConfiguration = [&]() {
+    workgroupSize = {subgroupSize, 1, 1};
+    workgroupTileSizes.resize(loopDepth, 0);
+    threadTileSizes.resize(loopDepth, 0);
+
+    // Initialize tiling along all partitioned loops with size 1.
+    for (int64_t loopIndex : partitionedLoops) {
+      workgroupTileSizes[loopIndex] = threadTileSizes[loopIndex] = 1;
+    }
+    // Override the innermost dimension to distribute to threads in a subgroup.
+    workgroupTileSizes.back() = subgroupSize;
+    threadTileSizes.back() = 1;
+  };
+
+  // Special case for non-linalg ops.
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp || linalgOp.getNumOutputs() != 1) {
+    auto pipeline =
+        IREE::Codegen::DispatchLoweringPassPipeline::SPIRVDistribute;
+
+    initConfiguration();
+    TileSizesListType tileSizes;
+    tileSizes.push_back(workgroupTileSizes);
+    tileSizes.push_back(threadTileSizes);
+
+    return setOpConfigAndEntryPointFnTranslation(funcOp, op, tileSizes, {},
+                                                 pipeline, workgroupSize);
+  }
+
+  // Common case for all linalg ops.
+
+  // The core idea is to distribute the partitioned loops to the workgroup
+  // dimensions. The goal is to fill up the GPU as much as possible, which means
+  // 1) distributing to as many threads as possible, and 2) avoid assigning too
+  // many threads to handle out-of-bound elements (thus idle).
+
+  SmallVector<TiledLoopInfo> tiledLoopInfo = getTiledLoopInfo(funcOp);
+  // The number of linalg implicit loops to partition and tiled loops
+  // surrounding the op should match. Otherwise, something is incorrect.
+  assert(partitionedLoops.size() == tiledLoopInfo.size());
+
+  // The upper bound for each implicit loop: 0 - untiled, negative - dynamic.
+  SmallVector<int64_t> loopBounds(loopDepth, 0);
+  // tiledLoopInfo uses the reverse order of partitionedLoops.
+  for (auto pair : llvm::zip(llvm::reverse(partitionedLoops), tiledLoopInfo)) {
+    unsigned loopIndex = std::get<0>(pair);
+    const TiledLoopInfo &loopInfo = std::get<1>(pair);
+    if (auto attr = loopInfo.ub.dyn_cast<Attribute>()) {
```

Comment:
Could replace this with some utility methods from here https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h

Reply:
Done